client/main.js

```diff
@@ -0,0 +1,470 @@
+    const repCycle = ReputationMiningCycle.at(addr);
+
+    let index = new BN("-1");
+    const round = new BN("0");
+    let submission = [];
+    while (submission[0] !== submittedHash) {
+      try {
+        index.iaddn(1);
+        submission = await repCycle.disputeRounds(round.toString(), index.toString()); // eslint-disable-line no-await-in-loop
+      } catch (err) {
+        round.iaddn(1);
+        index = new BN("-1");
+      }
+    }
+    return [round, index];
+  }
+
+  async respondToBinarySearchForChallenge() {
+    const [round, index] = await this.getMySubmissionRoundAndIndex();
+    const addr = await this.colonyNetwork.getReputationMiningCycle.call();
+    const repCycle = ReputationMiningCycle.at(addr);
+    let submission = await repCycle.disputeRounds(round.toString(), index.toString());
+    const targetNode = new BN(
+      submission[8]
+        .plus(submission[9])
+        .div(2)
+        .floor()
+        .toString()
+    );
+    const intermediateReputationHash = this.justificationHashes[`0x${targetNode.toString(16, 64)}`].jhLeafValue;
+    const [branchMask, siblings] = await this.justificationTree.getProof(`0x${targetNode.toString(16, 64)}`);
+
+    await repCycle.binarySearchForChallenge(round.toString(), index.toString(), intermediateReputationHash, branchMask, siblings, {
+      from: this.minerAddress,
+      gas: 1000000
+    });
+    submission = await repCycle.disputeRounds(round.toString(), index.toString());
+  }
+
+  async respondToChallenge() {
+    const [round, index] = await this.getMySubmissionRoundAndIndex();
+    const addr = await this.colonyNetwork.getReputationMiningCycle.call();
+    const repCycle = ReputationMiningCycle.at(addr);
+    const submission = await repCycle.disputeRounds(round.toString(), index.toString());
+    // console.log(submission);
+    const firstDisagreeIdx = new BN(submission[8].toString());
+    const lastAgreeIdx = firstDisagreeIdx.subn(1);
+    // console.log('getReputationUPdateLogEntry', lastAgreeIdx);
+    const logEntry = await this.colonyNetwork.getReputationUpdateLogEntry(lastAgreeIdx.toString(), false);
+    // console.log('getReputationUPdateLogEntry done');
+    const colonyAddress = logEntry[3];
+    const skillId = logEntry[2];
+    const userAddress = logEntry[0];
+    const reputationKey = `0x${new BN(colonyAddress.slice(2), 16).toString(16, 40)}${new BN(skillId.toString()).toString(16, 64)}${new BN(
+      userAddress.slice(2),
+      16
+    ).toString(16, 40)}`;
+    // console.log('get justification tree');
+    const [agreeStateBranchMask, agreeStateSiblings] = await this.justificationTree.getProof(`0x${lastAgreeIdx.toString(16, 64)}`);
+    const [disagreeStateBranchMask, disagreeStateSiblings] = await this.justificationTree.getProof(`0x${firstDisagreeIdx.toString(16, 64)}`);
+    // console.log('get justification tree done');
+
+    // These comments can help with debugging. This implied root is the intermediate root hash that is implied
```

Comment:
Maybe consider adding logging to console with various verbosity options for debugging. Something like https://github.com/visionmedia/debug should help